activesupport/lib/active_support/cache/key.rb

```diff
@@ -0,0 +1,153 @@
+    # `<<` method.
+    #
+    # Example:
+    #
+    #   key = Key.new
+    #   key << "foo"
+    #   key << "bar"
+    #   key.cache_key
+    #   # => "foo/bar"
+    #
+    # Multiple key entries generate the same key regardless of how
+    # they are added. I.e. if they're added via multiple arrays
+    # or directly via the `<<` operator.
+    #
+    # Example:
+    #
+    #   k1 = Key.new("foo")
+    #   k1 << "bar"
+    #   k2 = Key.new(["foo", "bar"])
+    #   k1.cache_key == k2.cache_key
+    #   # => true
+    #   k3 = Key.new(["foo", ["bar"]])
+    #   k2.cache_key == k3.cache_key
+    #   # => true
+    #
+    # It is faster to directly add objects to the
+    # Key object via `<<` than to generate
+    # an intermediate array of objects.
+    #
+    # The class supports strings, arrays, hashes,
+    # objects with a `cache_key` method and anything
+    # that can be implicitly cast to a string:
+    #
+    # Examples:
+    #
+    #   Key.new("foo").cache_key
+    #   # => "foo"
+    #
+    #   Key.new(["foo", "bar"]).cache_key
+    #   # => "foo/bar"
+    #
+    #   Key.new({ bar: 1, foo: 2 }).cache_key
+    #   # => "bar=1/foo=2"
+    #
+    #   klass = Class.new { def cache_key; "foo"; end }
+    #   Key.new(klass.new).cache_key
+    #   # => "foo"
+    #
+    # In addition to generating a `cache_key`,
+    # a `cache_version` is also generated if the object
+    # passed in responds to `cache_version` or if it contains
+    # an element that responds to `cache_version`.
+    #
+    # The purpose of keeping this value seperate from
+    # `cache_key` is to support the use of "cache versioning"
+    # where the same key is recycled but the version of the object
+    # being stored is kept directly in the cache.
+    #
+    # Example:
+    #
+    #   klass = Class.new { def cache_version; "foo"; end }
+    #   Key.new(klass.new).cache_version
+    #   # => "foo"
+    #
+    #   klass = Class.new { def cache_version; "foo"; end }
+    #   Key.new([klass.new]).cache_version
+    #   # => "foo"
+    #
+    #   klass = Class.new { def cache_version; "foo"; end }
+    #   Key.new({foo: klass.new}).cache_version
+    #   # => "foo"
+    class Key # :nodoc:
+      DEFAULT_KEY = Object.new
+      CACHE_METHOD_OBJ = ActiveSupport::Cache::Store.new
+      @@has_rails_cache = :unknown
+
+      def initialize(key = DEFAULT_KEY)
+        @cache_key = nil
+        @cache_version = nil
+        @key_parts = []
+        @cache_method_obj = cache_method_obj
+
+        self << key unless key == DEFAULT_KEY
+      end
+
+      def cache_key
+        @cache_key ||= @cache_method_obj.send(:expanded_key, @key_parts)
+      end
+
+      def cache_version
+        @cache_version ||= @cache_method_obj.send(:expanded_version, @key_parts)
+
+        if @cache_version == "/"
+          nil
+        else
+          @cache_version
+        end
+      end
+
+      def cache_key_with_version
+        @cache_method_obj.send(:retrieve_cache_key, @key_parts)
```

Comment:
Hm, the `send`s in here bother me, but I think we can live with them if the future design idea is to flip the relationship such that a cache store really consults the key for `retrieve_cache_key` etc.

Reply:
> if the future design idea is to flip the relationship such that a cache store really consults the key for retrieve_cache_key etc.

Inverting that relationship is the core idea with #36310 where the key logic would live in the object rather than in external recursive methods (i.e. `normalize_key` and `normalize_version`). It might be worth re-visiting it now if the eventual goal would be to get there. I could make it less invasive and restrict the key generation to only view collection caching. 

> Hm, the sends in here bother me

I agree that the sends here are unnerving. The interface provided by Cache::Store is a bit odd since most people who implement a cache store inherit from this store meaning they can freely use these methods even though they're written as being private. I could make them public (because in a very real way they're already public today) however that wouldn't fix any issues with other cache stores that have implemented their own methods and decided to keep them private.

This also does highlight probably the most "unsafe" thing about this PR versus #36310, it assumes any cache store provided has these 3 methods on them. To my knowledge, the vast majority of people use the Rails provided cache stores and some use `:dalli`. I'm not sure if there are any other cache stores to consider.